cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,415 @@
+        } else if (tag.Equals(TEXT("calibration"))) {
+            ASSERT(!has_calibration_node);
+            has_calibration_node = true;
+            for (auto& attr_node : node->GetAttributes()) {
+                if (attr_node.GetTag() == "rising") {
+                    ASSERT(joint_desc.calibration_type_ == CalibrationType::Invalid);
+                    joint_desc.calibration_type_ = CalibrationType::Rising;
+                    joint_desc.rising_ = FCString::Atof(*attr_node.GetValue());
+                } else if (attr_node.GetTag() == "falling") {
+                    ASSERT(joint_desc.calibration_type_ == CalibrationType::Invalid);
+                    joint_desc.calibration_type_ = CalibrationType::Falling;
+                    joint_desc.falling_ = FCString::Atof(*attr_node.GetValue());
+                }
+            }
+        } else if (tag.Equals(TEXT("dynamics"))) {
+            ASSERT(!has_dynamics_node);
+            has_dynamics_node = true;
+            joint_desc.damping_ = FCString::Atof(*node->GetAttribute(TEXT("damping")));
+            joint_desc.friction_ = FCString::Atof(*node->GetAttribute(TEXT("friction")));
+        } else if (tag.Equals(TEXT("limit"))) {
+            ASSERT(!has_limit_node);
+            has_limit_node = true;
+            joint_desc.lower_ = FCString::Atof(*node->GetAttribute(TEXT("lower")));
+            joint_desc.upper_ = FCString::Atof(*node->GetAttribute(TEXT("upper")));
+            joint_desc.velocity_ = FCString::Atof(*node->GetAttribute(TEXT("velocity")));
+            joint_desc.effort_ = FCString::Atof(*node->GetAttribute(TEXT("effort")));
+        } else if (tag.Equals(TEXT("mimic"))) {
+            ASSERT(!has_mimic_node);
+            has_mimic_node = true;
+            joint_desc.joint_ = TCHAR_TO_UTF8(*node->GetAttribute(TEXT("joint")));
+            joint_desc.multiplier_ = FCString::Atof(*node->GetAttribute(TEXT("multiplier")));
+            joint_desc.offset_ = FCString::Atof(*node->GetAttribute(TEXT("offset")));
+        } else if (tag.Equals(TEXT("safety_controller"))) {
+            ASSERT(!has_safety_controller_node);
+            has_safety_controller_node = true;
+            joint_desc.soft_lower_limit_ = FCString::Atof(*node->GetAttribute(TEXT("soft_lower_limit")));
+            joint_desc.soft_upper_limit_ = FCString::Atof(*node->GetAttribute(TEXT("soft_upper_limit")));
+            joint_desc.k_position_ = FCString::Atof(*node->GetAttribute(TEXT("k_position")));
+            joint_desc.k_velocity_ = FCString::Atof(*node->GetAttribute(TEXT("k_velocity")));
+        }
+    }
+
+    ASSERT(has_parent_node);
+    ASSERT(has_child_node);
+    // revolute and prismatic joints require limit
+    ASSERT(!(joint_desc.type_ == UrdfJointType::Revolute || joint_desc.type_ == UrdfJointType::Prismatic) || has_limit_node);
+
+    return joint_desc;
+}
+
+UrdfRobotDesc UrdfParser::parseRobotNode(FXmlNode* robot_node)
```

Comment:
Need to parse the top-level `name` in the robot node in this function.

Reply:
fixed